lib/board.ex

```
@@ -0,0 +1,40 @@
+defmodule Board do
+  defstruct [:one, :two, :three, :four, :five, :six, :seven, :eight, :nine]
+
+  def empty() do
+    %Board{
+      :one => "1",
+      :two => "2",
+      :three => "3",
+      :four => "4",
+      :five => "5",
+      :six => "6",
+      :seven => "7",
+      :eight => "8",
+      :nine => "9"
+    }
+  end
+
+  def place_mark(board = %Board{}, key, value) do
+    validate_value!(value)
```

Does this mean that if we are playing a game and one of us makes a mistake then the whole program will crash and we lose our progress?

---

Currently, yes.  This is a work in progress.  Eventually, the user interface will prevent such things from occurring.  Like most things, I am just guessing at how to implement it.  However I did read that Erlang has a 'let it crash philosophy' and this does seem to follow that mindset.

---

That is true. https://stackoverflow.com/questions/676266/how-to-explain-erlangs-supervision-trees

You shouldn't need a supervision tree for user input. And as a general rule I steer away from throwing / catching whenever possible (in favor of more functional approaches).